### PR TITLE
feat: add youtube notifications

### PR DESCRIPTION
**Pull request by Tylertron1998**
todo:
- [x] generally clean up code (refactor things into testable methods)
- [x] tests
- [x] add more logging (things can and will go wrong, but the more logs = more info for why)
- [x] write up a document on how this all works (for future reference, any contributors won't have to dig through 29348 different spec documents, all of which outdated, and browse forums just to understand the first 4 lines)

### CODE REVIEW

File: services/Skyra.Notifications/Services/YoutubeService.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+					continue;
+				}
+
+				_logger.LogInformation("Sending notification {@Notification}", notification);
+
+				await using var database = new SkyraDatabase(new SkyraDbContext(), new NullLogger<SkyraDatabase>());
+
+				var notificationSubscription = await database.GetSubscriptionAsync(notification.ChannelId);
+
+				var channels = new NotificationChannel[notificationSubscription.Value.GuildIds.Length];
```
Comment (Tylertron1998) [inline services/Skyra.Notifications/Services/YoutubeService.cs]: currently not checking here if no guilds are subscribed (though technically it's handled by the for loop). Will need to check and remove the entry from the database all together

File: services/Skyra.Database/Models/Entities/Guild.cs [hunk 2]
```diff
@@ -743,6 +743,14 @@ public class Guild
 		[Column("notifications.streams.twitch.streamers", TypeName = "jsonb")]
 		public string NotificationsStreamsTwitchStreamers { get; set; } = null!;
 
+		[Required]
+		[Column("notifications.youtube.channel", TypeName = "character varying(19)")]
+		public string YoutubeNotificationChannel { get; set; } = null!;
+
+		[Required]
+		[Column("notifications.youtube.message")]
+		public string YoutubeNotificationMessage { get; set; } = null!;
```
Comment (kyranet) [inline services/Skyra.Database/Models/Entities/Guild.cs]: Those can be nullable:
```suggestion
		[Column("notifications.youtube.channel", TypeName = "character varying(19)")]
		public ?string YoutubeNotificationChannel { get; set; };

		[Column("notifications.youtube.message")]
		public ?string YoutubeNotificationMessage { get; set; };
```

File: services/Skyra.Database/Models/Entities/YoutubeSubscription.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+	[Table("youtube_subscription")]
+	public class YoutubeSubscription
+	{
+		[Key]
+		[Column("id")]
+		[StringLength(24)]
+		public string Id { get; set; }
+
+		[Column("already_seen_ids", TypeName = "character varying(11)[]")]
+		public string[] AlreadySeenIds { get; set; }
```
Comment (kyranet) [inline services/Skyra.Database/Models/Entities/YoutubeSubscription.cs]: Required, assign to empty array or `null!`

File: services/Skyra.Database/Models/Entities/YoutubeSubscription.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+		[Column("expires_at")]
+		public DateTime ExpiresAt { get; set; }
+
+		[Required]
+		[Column("guild_ids", TypeName = "character varying(19)[]")]
+		public string[] GuildIds { get; set; }
```
Comment (kyranet) [inline services/Skyra.Database/Models/Entities/YoutubeSubscription.cs]: Assign to empty array or `null!`

File: services/Skyra.Database/Models/Entities/YoutubeSubscription.cs (hunk 3 above)
Comment (kyranet) [reply services/Skyra.Database/Models/Entities/YoutubeSubscription.cs]: Required, assign to empty array or `null!`

File: services/Skyra.Database/Models/Entities/YoutubeSubscription.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+		[Column("channel_title")]
+		public string ChannelTitle { get; set; }
```
Comment (kyranet) [inline services/Skyra.Database/Models/Entities/YoutubeSubscription.cs]: Required, assign to `null!`

File: services/Skyra.Database/SkyraDbContext.cs [hunk 6]
```diff
@@ -44,6 +44,8 @@ public SkyraDbContext(DbContextOptions<SkyraDbContext> options)
 		public virtual DbSet<UserProfile> UserProfiles { get; set; }
 		public virtual DbSet<UserSpousesUser> UserSpousesUsers { get; set; }
 
+		public virtual DbSet<YoutubeSubscription> YoutubeSubscriptions { get; set; }
```
Comment (kyranet) [inline services/Skyra.Database/SkyraDbContext.cs]:
```suggestion
		public virtual DbSet<YoutubeSubscription> YoutubeSubscriptions { get; set; }
```

File: services/Skyra.Notifications/Controllers/PubSubResponseController.cs [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+		public async Task Authenticate()
+		{
+			var topic = Request.Query["hub.topic"];
+			var mode = Request.Query["hub.mode"];
+			var challenge = Request.Query["hub.challenge"];
+			var queryParams = HttpUtility.ParseQueryString(new Uri(topic).Query);
+
+			var channelId = queryParams["channel_id"];
+			var isSubscription = mode == "subscribe";
+			// TODO: why is this not logging the lease time correctly?
```
Comment (kyranet) [inline services/Skyra.Notifications/Controllers/PubSubResponseController.cs:45]: Resolve TODO

File: services/Skyra.Notifications/Controllers/PubSubResponseController.cs [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+			_logger.LogInformation("Received Authentication request with challenge {Challenge} for topic {Topic} ({Mode})",
+				challenge, channelId, mode);
+
+			if (!_cache.GetRequest(channelId, isSubscription))
```
Comment (kyranet) [inline services/Skyra.Notifications/Controllers/PubSubResponseController.cs:49]: channelId is nullable, please handle it

File: services/Skyra.Notifications/Controllers/PubSubResponseController.cs [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+			if (ids.Contains(videoId))
+			{
+				// we have already posted this video, so let's ignore it, as youtube likes to send duplicates
+				_logger.LogInformation("Ignoring event with video ID of {VideoId} because it's already been seen in the database", videoId);
+				return Ok();
+			}
+
+			// at this point, we can be pretty sure that we've covered all bases.
+			// thank you, google, for making me suffer.
```
Comment (kyranet) [inline services/Skyra.Notifications/Controllers/PubSubResponseController.cs:103]: Same

File: services/Skyra.Notifications/Models/Notification.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+
+namespace Skyra.Notifications.Models
+{
+	public class Notification
+	{
+		public string VideoId { get; set; }
```
Comment (kyranet) [inline services/Skyra.Notifications/Models/Notification.cs]: There's a bunch of warnings in this class

File: services/Skyra.Notifications/Program.cs [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+{
+	public class Program
+	{
+		public static async Task Main(string[] args)
+		{
+			AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2UnencryptedSupport", true);
+			AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2Support", true);
+			var host = CreateHostBuilder(args).Build();
+			var manager = host.Services.GetService<SubscriptionManager>();
+			await manager.StartAsync();
```
Comment (kyranet) [inline services/Skyra.Notifications/Program.cs:19]: Resolve this warning (manager is possibly `null`)

File: services/Skyra.Notifications/Services/YoutubeService.cs (hunk 1 above)
Comment (kyranet) [reply services/Skyra.Notifications/Services/YoutubeService.cs]: Is this resolved yet?

File: services/Skyra.Notifications/SubscriptionManager.cs [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+		private readonly PubSubClient _pubSubClient;
+		private readonly Timer _resubTimer;
+		private HttpClient _httpClient;
+
+		private JsonSerializerOptions _jsonSerializerOptions = new JsonSerializerOptions
+		{
+			PropertyNamingPolicy = JsonNamingPolicy.CamelCase
+		};
+
+		private Dictionary<string, DateTime> _resubscribeTimes;
```
Comment (kyranet) [inline services/Skyra.Notifications/SubscriptionManager.cs]: Initialize this

File: services/Skyra.UnitTests/Notifications/RequestCacheTests.cs [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+			// arrange
+
+			var cache = new RequestCache(new NullLogger<RequestCache>());
+
+			// act
+
+			cache.AddRequest(id, isSubscription);
+			var _ = cache.GetRequest(id, isSubscription);
+			var result = cache.GetRequest(id, isSubscription);
+			// assert
```
Comment (kyranet) [inline services/Skyra.UnitTests/Notifications/RequestCacheTests.cs]:
```suggestion
			var result = cache.GetRequest(id, isSubscription);

			// assert
```

File: services/Skyra.UnitTests/Notifications/RequestCacheTests.cs [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+			// arrange
+
+			var cache = new RequestCache(new NullLogger<RequestCache>());
+
+			// act
+
+			cache.AddRequest(id, isSubscription);
+			var _ = cache.GetRequest(id, isSubscription, false);
+			var result = cache.GetRequest(id, isSubscription);
+			// assert
```
Comment (kyranet) [inline services/Skyra.UnitTests/Notifications/RequestCacheTests.cs]:
```suggestion
			var result = cache.GetRequest(id, isSubscription);

			// assert
```

**Review by kyranet — CHANGES_REQUESTED** [review]

File: services/Skyra.Notifications/SubscriptionManager.cs (hunk 12 above)
Comment (Tylertron1998) [reply services/Skyra.Notifications/SubscriptionManager.cs]: [it is](https://github.com/skyra-project/skyra/pull/1910/files/e9dd02d1dc93bc5b5bd9acb3cb54cf7fc5c33e2e#diff-b92397fa1a889e64810aa7d57ee54ebb60803ec3040aebd51102cabddea12856R71)

**Comment by kyranet** [comment]
Closing as this has been superseded by [Arkadia](https://github.com/skyra-project/arkadia).

**Pull request closed without merging**